src/TriggerActivityMakerBundleN.cpp

```diff
@@ -0,0 +1,104 @@
+ * received with this code.
+ */
+
+#include "triggeralgs/BundleN/TriggerActivityMakerBundleN.hpp"
+
+#include "TRACE/trace.h"
+#define TRACE_NAME "TriggerActivityMakerBundleNPlugin"
+
+namespace triggeralgs {
+
+bool TriggerActivityMakerBundleN::bundle_condition() {
+  return m_current_ta.inputs.size() == m_bundle_num;
+}
+
+void
+TriggerActivityMakerBundleN::operator()(const TriggerPrimitive& input_tp, std::vector<TriggerActivity>& output_tas)
+{
+  // Expect that TPs are inherently time ordered.
+  m_current_ta.inputs.push_back(input_tp);
+
+  if (bundle_condition()) {
+    // Using the first TA as reference.
+    TriggerPrimitive first_tp = m_current_ta.inputs.front();
+    TriggerPrimitive last_tp = m_current_ta.inputs.back();
+
+    m_current_ta.channel_start = first_tp.channel;
+    m_current_ta.channel_end = last_tp.channel;
+
+    m_current_ta.time_start = first_tp.time_start;
+    m_current_ta.time_end = last_tp.time_start;
+
+    m_current_ta.detid = first_tp.detid;
+
+    m_current_ta.algorithm = TriggerActivity::Algorithm::kBundle;
+    m_current_ta.type = TriggerActivity::Type::kTPC;
+
+    m_current_ta.adc_peak = 0;
+    for (const TriggerPrimitive& tp : m_current_ta.inputs) {
+      m_current_ta.adc_integral += tp.adc_integral;
+      if (tp.adc_peak <= m_current_ta.adc_peak) continue;
+      m_current_ta.adc_peak = tp.adc_peak;
+      m_current_ta.channel_peak = tp.channel;
+      m_current_ta.time_peak = tp.time_peak;
+    }
+    m_current_ta.time_activity = m_current_ta.time_peak;
+
+    output_tas.push_back(m_current_ta);
+
+    // Reset the current.
+    m_current_ta = TriggerActivity();
+  }
+
+  // Should never reach this step. In this case, send it out.
+  if (m_current_ta.inputs.size() > m_bundle_num) {
+    TLOG(TLVL_DEBUG_1) << "Emitting large BundleN TriggerActivity with " << m_current_ta.inputs.size() << " TPs.";
+
+    // Using the first TA as reference.
+    TriggerPrimitive first_tp = m_current_ta.inputs.front();
+    TriggerPrimitive last_tp = m_current_ta.inputs.back();
+
+    m_current_ta.channel_start = first_tp.channel;
+    m_current_ta.channel_end = last_tp.channel;
+
+    m_current_ta.time_start = first_tp.time_start;
+    m_current_ta.time_end = last_tp.time_start;
+
+    m_current_ta.detid = first_tp.detid;
+
+    m_current_ta.algorithm = TriggerActivity::Algorithm::kBundle;
+    m_current_ta.type = TriggerActivity::Type::kTPC;
+
+    m_current_ta.adc_peak = 0;
+    for (const TriggerPrimitive& tp : m_current_ta.inputs) {
+      m_current_ta.adc_integral += tp.adc_integral;
+      if (tp.adc_peak <= m_current_ta.adc_peak) continue;
+      m_current_ta.adc_peak = tp.adc_peak;
+      m_current_ta.channel_peak = tp.channel;
+      m_current_ta.time_peak = tp.time_peak;
+    }
+    m_current_ta.time_activity = m_current_ta.time_peak;
+
+    output_tas.push_back(m_current_ta);
+
+    // Reset the current.
+    m_current_ta = TriggerActivity();
+  }
```

Comment:
Looks identical to the if function above, maybe worth abstracting out into a function like in other makers...
Also, if `m_current_ta.inputs.size() > m_bundle_num` is really met, then shouldn't the difference be taken into the account when constructing the TA? I.e. the last few TPs wouldn't be taken into the account, and would be pushed into the new TA. Not that it matters, as you said, we should never reach this place...

Reply:
In the ideal case, probably. However, I imagine if the system is having problems at this level, this won't really be the top priority. It never ran into this block while testing, so that's at least a good sign.

I'll make a function that does the `set_ta_attributes` as in the ChannelDistance algorithm. It may be useful for such a function to be in the base `TriggerXMaker` class.

Reply:
I agree about having such function in the `TriggerXMaker` class -- but let's wait with that one till v5.